Use `string` type for Document id in tests

franmomu opened the pull request:
Ref: https://github.com/ruflin/Elastica/issues/2079#issuecomment-1160163215

~This PR allows `int` as possible type for `id` when creating `Document` and also in `AbstractScript`, `Script` and `ParentId`.~

~The idea of this PR is to be BC, but then we should change `AbstractUpdateAction::setId()`, `Index::createDocument()` and maybe some other that would mean BC break, for them we could maybe leave it as it is, ignore the error in PHPStan and add some kind of reminder to change it in `8.x`.~

~BTW with this change errors in PHPStan level 5 go from 624 to 81.~

This is just a proposal, we could also ignore those errors in PHPStan. I'm open to any alternative.

**Update**: This PR updates our test suite to use `string` type for `Document` ids.

ruflin commented on src/AbstractUpdateAction.php:
```diff
@@ -26,6 +28,8 @@ public function setId(?string $id = null): self
     }
 
     /**
+     * TODO: Allow also "int" as return type.
```
I think here it should always be an `string`. Even if you set is a `int` it is converted to a string. That you can set it as an int is more a convenience for the dev but not for internal representation.

ruflin reviewed (commented):
When I made the comment in https://github.com/ruflin/Elastica/issues/2079#issuecomment-1160163215 it all sounded so simple but now looking at the PR things are a bit more complicated. For example, should we really "loosen" strict mappings that we have? It seems it worked so far and users were happy with it. 

Would the following be possible?

* Keep strict mappings where we already had it
* Update support for `int` in the docs where it was missing
* Update tests (if needed)?

Not sure I get to this PR again before I'm out, @thePanz would be great if you could have a look.

franmomu commented:
I changed it as draft until we see it is fine.

I've been checking the dependent packages in https://packagist.org/packages/ruflin/elastica/dependents?order_by=downloads (the most used ones) and what I've seen so far is that the ones using `Document`, they use a proper `string` as id.

> It seems it worked so far and users were happy with it.

I agree with this, since it has been typed since version 7 (and in the PHPDoc before), I think it is safeish to say that users have updated their code.

> Would the following be possible?
> 
> * Keep strict mappings where we already had it

Done.

> * Update support for `int` in the docs where it was missing

do you mean add `int` so it looks like:

```php
/**
 * Creates a new document.
 *
 * @param int|string|null  $id    The document ID, if null it will be created
 * @param array|string $data  Data array
 * @param Index|string $index Index name
 */
public function __construct(?string $id = null, $data = [], $index = '')
```
and when support for php < 8 is dropped, change it to `string|int $id`? and internally cast it always to `string`
> * Update tests (if needed)?
> 

I've updated most of the test (using regular expressions) to see how it would look like if we just keep `string`.

thePanz commented:
I would opt to keep our library accept and return the type used in ES.
It would feel weird to index a product with `int 123` and later load it, to have `string "123"` in the response.

I am more keen to keep the strict `string` type.

For this PR it means:
- code wise it is good, all calls have been casted to `string`
- remove the new lines from the cangelog :)

WDYT?

franmomu commented:
Yeah, the initial issue here was about [tests not being in sync with the types defined in `src` because of not using strict types](https://github.com/ruflin/Elastica/issues/2079#issue-1276007828).

This change should be fine and we'll see about adding strict types in the next major version.
